go/workflow.go

```diff
@@ -0,0 +1,333 @@
+	selector.AddReceive(workflow.GetSignalChannel(ctx, SignalEnsureMinimumStatus),
+		func(c workflow.ReceiveChannel, _ bool) {
+			errSignal = signalEnsureMinimumStatus(ctx, c, &customer)
+		})
+
+	// signal handler for canceling account
+	selector.AddReceive(workflow.GetSignalChannel(ctx, SignalCancelAccount),
+		func(c workflow.ReceiveChannel, _ bool) {
+			errSignal = signalCancelAccount(ctx, c, &customer)
+		})
+
+	// handle Temporal Server cancellation requests
+	selector.AddReceive(ctx.Done(),
+		func(c workflow.ReceiveChannel, _ bool) {
+			c.Receive(ctx, nil)
+			logger.Info("Cancellation requested.")
+			workflowCanceled = true
+		})
+
+	// query handler for status level, etc
+	err = workflow.SetQueryHandler(ctx, QueryGetStatus,
+		func() (GetStatusResponse, error) {
+			return queryGetStatus(ctx, customer)
+		})
+	if err != nil {
+		return fmt.Errorf("unable to register '%v' query handler: %w", QueryGetStatus, err)
+	}
+
+	// query handler for guest list
+	err = workflow.SetQueryHandler(ctx, QueryGetGuests,
+		func() ([]string, error) {
+			return queryGetGuests(ctx, customer)
+		})
+	if err != nil {
+		return fmt.Errorf("unable to register '%v' query handler: %w", QueryGetGuests, err)
+	}
+
+	// Block on everything. Continue-As-New on history length; size of activities in this workflow are small enough
+	// that we'll hit the length thresholds well before any size threshold.
+	logger.Info("Waiting for new signals")
+	for customer.AccountActive && info.GetCurrentHistoryLength() < EventsThreshold && !workflowCanceled {
+		selector.Select(ctx)
+
+		if errSignal != nil {
+			logger.Error("Unrecoverable error in handling a signal.", "Error", errSignal)
+			return errSignal
+		}
+	}
+
+	// here because of events threshold, but account still active? Continue-As-New
+	if customer.AccountActive && !workflowCanceled {
+		return workflow.NewContinueAsNewError(ctx, customer)
+	}
+
+	logger.Info("Loyalty workflow completed.", "Customer", customer, "WorkflowCanceled", workflowCanceled)
+	return nil
+}
+
+// CustomerWorkflowID generates a Workflow ID based on the given customer ID.
+func CustomerWorkflowID(customerID string) string {
+	return fmt.Sprintf("customer-%v", customerID)
+}
+
+func validateCustomerInfo(customer *CustomerInfo) {
+	if customer.StatusLevel == nil {
+		customer.StatusLevel = StatusLevelForPoints(customer.LoyaltyPoints)
+	}
+	if len(customer.Guests) == 0 {
+		customer.Guests = make([]string, customer.StatusLevel.GuestsAllowed)
+	}
+}
+
+func addGuestToCustomer(customer *CustomerInfo, guestID string) error {
+	guestExists := false
+	for _, g := range customer.Guests {
+		guestExists = g == guestID
+	}
+	// replace first empty slot, or append if none
+	if !guestExists {
+		added := false
+		for i, g := range customer.Guests {
+			if len(g) == 0 {
+				customer.Guests[i] = guestID
+				added = true
+				break
+			}
+		}
+		if !added {
+			customer.Guests = append(customer.Guests, guestID)
+		}
+	}
+
+	return nil
+}
```

Comment:
```suggestion
func (c *Customer) addGuest(guestID string) {
	// Add if not there
	for _, g := range c.Guests {
		if g == guestID {
			return
		}
	}
	c.Guests = append(c.Guests, guestID)
}
```

This is the more Go-like way (once you remove the setting of the `Guests` to an empty pre-sized slice). Also, seems clearer to make this a method on the customer.